### ISSUE TITLE
Borrow for Intern with HashMap has peculiar behavior

### ISSUE THREAD

**Issue by stuhood** [droundy/internment#36]
I'm not sure if it qualifies as a bug or just a gotcha, but code like this will not work as expected:
```rust
let mut mapping: HashMap<Intern<String>, _> = HashMap::new();
let x = "x".to_string();
mapping.insert(Intern::new(x.clone()), 1);
// A lookup for an equal `&String` will fail, so this line panics.
mapping.get(&x).unwrap();
```

**Comment by droundy** [comment]
To be clear, I believe the lookup fails, and because you unwrap, it panics. This happens because the hash of an Intern differs from the hash of its contents.  That's due to the feature of an Intern that makes it faster to look up in a HashMap because you don't need to deference any pointers.

I think it's a gotcha rather than a bug, but probably be worth documenting.

**Comment by stuhood** [comment]
> That's due to the feature of an Intern that makes it faster to look up in a HashMap because you don't need to deference any pointers.

Mm, yea: that aspect makes sense. But it looks like [Borrow](https://doc.rust-lang.org/std/borrow/trait.Borrow.html)'s docs discuss the case.

> Can `CaseInsensitiveString` implement `Borrow<str>`? It certainly can provide a reference to a string slice via its contained owned string. But because its `Hash` implementation differs, it behaves differently from `str` and therefore must not, in fact, implement `Borrow<str>`. If it wants to allow others access to the underlying `str`, it can do that via AsRef<str> which doesn’t carry any extra requirements.

So... it seems like it might be a violation of `Borrow`'s contract to implement it for `Intern`?

**Comment by droundy** [comment]
That sounds correct, I guess I never read and retained Borrow's documentation.  Sounds like I should remove the Borrow implementation.

**Comment by Eh2406** [comment]
From the changelog this was fixed back in 0.7.0.
